### PR TITLE
Fix aggregations with multiple grouping sets and no input

### LINKED ISSUE #6066: Incorrect results for GROUPING SETS with multiple () groups
This query:

```sql
SELECT count(*) 
FROM (VALUES 1, 2, 3) t(x) 
WHERE false 
GROUP BY GROUPING SETS ((), ())
```

incorrectly produces no rows. It should produce two rows with "0" in each cell.

### PR DESCRIPTION
**Pull request by raghavsethi**
When an aggregation with an empty grouping set, cube or rollup
processed no input rows, no output was produced. This is incorrect, as
empty grouping sets are considered to be global aggregations and must
generate a single row even if no input was processed. This commit adds
the following:

- Detecting when there is an empty grouping set, and changing the plan
  to collapse to a single final aggregration operator.
- Changing HashAggregationOperator to generate default outputs when
  there are global aggregations and no input has been processed.
- Tests for several cases when no input is received.

Fixes #6066

### CODE REVIEW

**Comment by raghavsethi** [comment]
@martint this is hi-pri as it fixes a release blocker.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/AggregationNode.java [hunk 1]
```diff
@@ -167,6 +169,19 @@ public double getConfidence()
         return groupingSets;
     }
 
+    @JsonProperty("partitioningRequirement")
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/AggregationNode.java]: Why is this json-serializable?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/AggregationNode.java [hunk 2]
```diff
@@ -167,6 +169,19 @@ public double getConfidence()
...
+    public Set<Symbol> getPartitioningRequirement()
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/AggregationNode.java]: I'm not sure I like this being in the node itself. It's a concern for the optimizers that have to make partitioning decisions.

File: presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java [hunk 3]
```diff
@@ -252,6 +287,39 @@ public Page getOutput()
...
 
+    private Page getGlobalAggregationOutput()
+    {
+        List<Accumulator> accumulators = accumulatorFactories.stream()
+                .map(AccumulatorFactory::createAccumulator)
+                .collect(Collectors.toList());
+
+        PageBuilder output = new PageBuilder(types);
+
+        for (Integer globalAggregationGroupId : globalAggregationGroupIds) {
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java]: This can be `int`

Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java]: Also, I'd name the variable just `groupId`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java [hunk 4]
```diff
@@ -1906,18 +1913,21 @@ private PhysicalOperation planGroupByAggregation(AggregationNode node, PhysicalO
 
             Optional<Integer> hashChannel = node.getHashSymbol().map(channelGetter(source));
 
+            ImmutableMap<Symbol, Integer> mappings = outputMappings.build();
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java]: Use `Map` as the declared type

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 5]
```diff
@@ -1334,6 +1334,15 @@ public void testGroupingSets()
     }
 
     @Test
+    public void testGroupingSetsNoInput()
+            throws Exception
+    {
+        assertQuery("SELECT linenumber, suppkey, SUM(CAST(quantity AS BIGINT)) FROM lineitem WHERE quantity < 0 GROUP BY GROUPING SETS ((linenumber, suppkey), (suppkey))",
```
Comment (martint) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: Format like this for better readability:

```java
assertQuery(
        <<query1>>,
        <<query2>>)
```

Comment (martint) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: Also, It'd probably be easier to follow if you broke up the queries like this:

```
SELECT ....
FROM ....
WHERE ...
GROUP BY ....
```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/AggregationNode.java [hunk 6]
```diff
@@ -100,7 +100,8 @@ public AggregationNode(@JsonProperty("id") PlanNodeId id,
         for (Symbol mask : masks.keySet()) {
             checkArgument(aggregations.containsKey(mask), "mask does not match any aggregations");
         }
-        this.groupingSets = ImmutableList.copyOf(requireNonNull(groupingSets, "groupingSets is null"));
+        checkArgument(!requireNonNull(groupingSets, "groupingSets is null").isEmpty(), "grouping sets list cannot be empty");
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/AggregationNode.java]: This is hard to read. Break it up into individual statements:

```java
requireNonNull(groupingSets, "groupingSets is null");
checkArgument(!groupingSets.isEmpty(), "grouping sets list cannot be empty");
```

**Comment by martint** [comment]
Minor comments. Otherwise, looks good.

**Pull request closed without merging**